### PR TITLE
Yac: store states from future

### PR DESCRIPTION
**Pull request by lebdron**
### Description of the Change
Reduce transaction latency by storing votes from future and applying them when info about the corresponding round is available.

### Benefits
Improved performance

### Possible Drawbacks 
None?

### CODE REVIEW

File: irohad/consensus/yac/impl/yac.cpp [hunk 1]
```diff
@@ -85,11 +85,21 @@ namespace iroha {
                                  [](const auto &p) { return p->address(); }),
                        ", "));
 
+        std::vector<std::vector<VoteMessage>> states;
         std::unique_lock<std::mutex> lock(mutex_);
         cluster_order_ = order;
         alternative_order_ = std::move(alternative_order);
         round_ = hash.vote_round;
+        auto it = future_states_.lower_bound(round_);
```
Comment (kuvaldini) [inline irohad/consensus/yac/impl/yac.cpp:93]: Could you tell what does this block of code do in hunan lang?

Comment (kuvaldini) [reply irohad/consensus/yac/impl/yac.cpp:93]:
```suggestion
        // take states from current round if it was stored, erase older than current.
        auto it = future_states_.lower_bound(round_);
```
May be it is worth to create a named function or lambda for future readers generations.

File: irohad/consensus/yac/yac.hpp [hunk 2]
```diff
@@ -131,6 +132,7 @@ namespace iroha {
         std::shared_ptr<YacNetwork> network_;
         std::shared_ptr<YacCryptoProvider> crypto_;
         std::shared_ptr<Timer> timer_;
+        std::map<Round, std::vector<std::vector<VoteMessage>>> future_states_;
```
Comment (kuvaldini) [inline irohad/consensus/yac/yac.hpp:135]: I cannot find a plase where you fill this collection with data..

File: irohad/consensus/yac/impl/yac.cpp [hunk 3]
```diff
@@ -133,6 +143,13 @@ namespace iroha {
 
       void Yac::onState(std::vector<VoteMessage> state) {
         std::unique_lock<std::mutex> guard(mutex_);
+        auto &proposal_round = getRound(state);
+
+        if (proposal_round.block_round > round_.block_round) {
+          log_->info("Store state from future for {}", proposal_round);
+          auto &states = future_states_[proposal_round];
+          states.push_back(state);
```
Comment (kuvaldini) [inline irohad/consensus/yac/impl/yac.cpp:151]:
```suggestion
          future_states_[proposal_round].push_back(state);
```

**Review by kuvaldini — CHANGES_REQUESTED** [review]

**Comment by lebdron** [comment]
@kuvaldini @iceseer please check. You can use "hide whitespace changes".

File: irohad/consensus/impl/round.cpp [hunk 4]
```diff
@@ -12,38 +12,44 @@
...
-    std::string Round::toString() const {
-      return shared_model::detail::PrettyStringBuilder()
-          .init("Round")
-          .appendNamed("block", block_round)
-          .appendNamed("reject", reject_round)
-          .finalize();
-    }
-  }  // namespace consensus
-}  // namespace iroha
+namespace iroha::consensus {
```
Comment (kuvaldini) [inline irohad/consensus/impl/round.cpp:15]: This makes C++17 minimum required. Is it set in CMake? Is it documented?

File: irohad/consensus/yac/impl/yac.cpp [hunk 5]
```diff
@@ -24,336 +24,347 @@ auto &getRound(const std::vector<iroha::consensus::yac::VoteMessage> &state) {
...
+    alternative_order_ = std::move(alternative_order);
+    round_ = hash.vote_round;
+    // take states from current round if it was stored, erase older than current
+    auto it = future_states_.lower_bound(round_);
+    if (it->first == round_) {
+      state = std::move(it->second);
+      ++it;
+    }
+    future_states_.erase(future_states_.begin(), it);
+    lock.unlock();
```
Comment (kuvaldini) [inline irohad/consensus/yac/impl/yac.cpp:98]: Please use RAII

File: irohad/consensus/impl/round.cpp (hunk 4 above)
Comment (lebdron) [reply irohad/consensus/impl/round.cpp:15]: Yes https://github.com/hyperledger/iroha/blob/388c2d381363011238162f139130742f14d18422/CMakeLists.txt#L16

File: irohad/consensus/yac/impl/yac.cpp (hunk 5 above)
Comment (lebdron) [reply irohad/consensus/yac/impl/yac.cpp]: Fixed

**Comment by lebdron** [comment]
Ping @kuvaldini @iceseer

**Comment by lebdron** [comment]
Ping @iceseer @kuvaldini

**Review by iceseer — APPROVED** [review]

File: irohad/consensus/yac/impl/yac.cpp [hunk 6]
```diff
@@ -24,336 +24,349 @@ auto &getRound(const std::vector<iroha::consensus::yac::VoteMessage> &state) {
...
+
+  void Yac::onState(std::vector<VoteMessage> state) {
+    std::unique_lock<std::mutex> guard(mutex_);
+    auto &proposal_round = getRound(state);
+
+    if (not crypto_->verify(state)) {
+      log_->warn("Crypto verification failed for message. Votes: [{}]",
+                 boost::algorithm::join(
+                     state | boost::adaptors::transformed([](const auto &v) {
+                       return v.signature->toString();
```
Comment (kuvaldini) [inline irohad/consensus/yac/impl/yac.cpp:154]: `v.signature->toString()` size is about 64 or 128 symbols? What is the result string length for state of size 100?

How about redusing hash size?
```suggestion
                       return v.signature->toString().substr(0,7)+"..";
```

File: irohad/consensus/yac/impl/yac.cpp [hunk 7]
```diff
@@ -24,336 +24,349 @@ auto &getRound(const std::vector<iroha::consensus::yac::VoteMessage> &state) {
...
+
+  // ------|Hash gate|------
+
+  void Yac::vote(YacHash hash,
+                 ClusterOrdering order,
+                 boost::optional<ClusterOrdering> alternative_order) {
+    log_->info(
+        "Order for voting: [{}]",
+        boost::algorithm::join(
+            order.getPeers() | boost::adaptors::transformed([](const auto &p) {
```
Comment (kuvaldini) [inline irohad/consensus/yac/impl/yac.cpp:81]: Message length could be toooooo long if number of peers in network big. Let's think about compact log messages.

File: irohad/consensus/yac/impl/yac.cpp [hunk 8]
```diff
@@ -24,336 +24,349 @@ auto &getRound(const std::vector<iroha::consensus::yac::VoteMessage> &state) {
...
+                     }),
+                     ", "));
+      return;
+    }
+
+    // TODO lebdron: configurable high water mark (see section 4.4
+    // of https://doi.org/10.1145/571637.571640)
+    if (proposal_round.block_round > round_.block_round) {
+      log_->info("Store state from future for {}", proposal_round);
+      future_states_[proposal_round].insert(state.begin(), state.end());
```
Comment (kuvaldini) [inline irohad/consensus/yac/impl/yac.cpp:164]: Am I right if say "future_states_[proposal_round] must not exist". If so add `assert()`

**Review by kuvaldini — CHANGES_REQUESTED** [review]

**Comment by kuvaldini** [comment]
@lebdron Please rebase

**Pull request closed without merging**